numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -1333,55 +1326,18 @@ PyUFunc_MatmulTypeResolver(PyUFuncObject *ufunc,
             type_tup, out_dtypes) < 0) {
         return -1;
     }
-    if (PyArray_NDIM(operands[0]) >= 2 || PyArray_NDIM(operands[1]) >= 2) {
-#if defined(HAVE_CBLAS)
-        int typenum =  out_dtypes[2]->type_num;
-        if (  (NPY_DOUBLE == typenum || NPY_CDOUBLE == typenum ||
-                NPY_FLOAT == typenum || NPY_CFLOAT == typenum)) {
-            /*
-             * We are going to use BLAS
-             * make sure 2d and more arrays are contiguous
-             */
-            if (_bad_strides(operands[0]) || NOT_CONTIGUOUS(operands[0])) {
-                    PyObject *op = PyArray_NewCopy(operands[0], NPY_ANYORDER);
-
-                    if (op == NULL) {
-                       return -1;
-                    }
-                    Py_DECREF(operands[0]);
-                    operands[0] = (PyArrayObject *)op;
-            }
-            if (_bad_strides(operands[1]) || NOT_CONTIGUOUS(operands[1])) {
-                    PyObject *op = PyArray_NewCopy(operands[1], NPY_ANYORDER);
-
-                    if (op == NULL) {
-                        return -1;
-                    }
-                    Py_DECREF(operands[1]);
-                    operands[1] = (PyArrayObject *)op;
-            }
-        }
-#endif
-        if (operands[2] != NULL) {
-            npy_intp last_stride = PyArray_STRIDE(operands[2],
-                                                  PyArray_NDIM(operands[2]) - 1);
-            if (last_stride != PyArray_ITEMSIZE(operands[2])) {
-                PyErr_SetString(PyExc_ValueError,
-                    "output array is not acceptable (must be C-contiguous)");
-                return -1;
-            }
-            /*
-             * Use all the flags in PyUFunc_GeneralizedFunction
-             * except NPY_ITER_OVERLAP_ASSUME_ELEMENTWISE and
-             * NPY_ITER_WRITEONLY. Without NPY_ITER_OVERLAP_ASSUME_ELEMENTWISE
-             * nditer will properly check overlap between output and any input.
-             */
-            ufunc->op_flags[2] = NPY_ITER_WRITEONLY |
-                      NPY_ITER_UPDATEIFCOPY |
-                      NPY_ITER_ALIGNED |
-                      NPY_ITER_ALLOCATE |
-                      NPY_ITER_NO_BROADCAST;
-        }
+    if (operands[2] != NULL) {
+        /*
+         * Use all the flags in PyUFunc_GeneralizedFunction
+         * except NPY_ITER_OVERLAP_ASSUME_ELEMENTWISE and
+         * NPY_ITER_WRITEONLY. Without NPY_ITER_OVERLAP_ASSUME_ELEMENTWISE
+         * nditer will properly check overlap between output and any input.
+         */
+        ufunc->op_flags[2] = NPY_ITER_WRITEONLY |
```

Comment:
Why are we modifiying the global `np.matmul` object in the type resolution function? Is this thread-safe? Can we just set these properties at initialization?

Reply:
Perhaps this is the only way to get the behavior we want, but if we're going to do it this way, we should have a clear comment explaining why we have to modify it here, and noting it's probably a hack.

Reply:
Yes, it would be better to do this at ufunc initialization but there is no API for that.

Reply:
Why do you need an API for that? Can't we just do in `init_multiarray` or similar:

```
matmul_ufunc->op_flags[2] = ...
```

Reply:
yes, adopting. This makes the type selector function

Reply:
If possible, can you express the flags as an "or" of the default flags and whatever extras you're adding?

Reply:
I agree that it's sort of awkard to inject into the generated umath code.